### PR TITLE
feat(plugin-cc): Implementation for the Agent Status Feature

### PR DESCRIPTION
**Pull request by pagour98**
# COMPLETES #<[SPARK-558614](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-558614)>

## This pull request addresses

The user should be able to set Agent Status.

API:- https://api.wxcc-us1.cisco.com/v1/agents/session/state

Description:

The user should be able to set Agent Status by making a PUT request to the above API

## by making the following changes

1. Added UI and called relevant method inside Samples Page.
2. Introduced new functions inside cc.ts file
3. Introduced new functions inside AgentConfig file
4. Introduced new function inside AgentConfigService file

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

I have tested my changes through the CC samples app using the UI changes that I have add. I have tested the Post with valid and invalid request.

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-cc/src/features/Agent.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+      return {
+        data: loginResponse,
+      };
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
+
+  public async setAgentStatus(data: StateChange) {
```
Comment (arun3528) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: i feel there is no value add of this function , we could simple reexport await this.agentService.setAgentStatus from this file 

and this file could be converted to a helper file if there is a need or else we could simply use consume it directly in cc.ts

File: packages/@webex/plugin-cc/src/features/Agent.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+export default class Agent {
+  private webex: WebexSDK;
+  private agentService: AgentService;
+  private httpRequest: HttpRequest;
+
+  constructor(webex: WebexSDK, httpRequest: HttpRequest) {
+    this.webex = webex;
+    this.httpRequest = httpRequest;
+    this.agentService = new AgentService(webex, this.httpRequest);
```
Comment (arun3528) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: this can be moved to cc.ts as other modules would like task can reuse those

Comment (arun3528) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: and its a sigleton so you just import in places where you need them

File: packages/@webex/plugin-cc/src/features/Agent.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    try {
+      const loginResponse = await this.agentService.stationLogin({
+        dialNumber: data.dialNumber,
+        teamId: data.teamId,
+        deviceType: data.loginOption,
+        isExtension: data.loginOption === LoginOption.EXTENSION,
+        deviceId: this.getDeviceId(data.loginOption, data.dialNumber),
+        roles: [AGENT],
+      });
+      this.webex.logger.log('LOGIN API SUCCESS');
```
Comment (arun3528) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: this should use the logger proxy

**Review by arun3528 — CHANGES_REQUESTED** [review]

File: packages/@webex/plugin-cc/src/services/config/types.ts [hunk 4]
```diff
@@ -113,6 +113,13 @@ export type AgentLogin = {
   loginOption: LoginOption;
 };
 
+export interface StateChange {
```
Comment (rsarika) [inline packages/@webex/plugin-cc/src/services/config/types.ts:110]: it should be type.

File: packages/@webex/plugin-cc/src/services/config/types.ts [hunk 5]
```diff
@@ -113,6 +113,13 @@ export type AgentLogin = {
...
+  state: string;
+  auxCodeId: string;
+  lastStateChangeReason?: string;
+  agentId?: string;
```
Comment (rsarika) [inline packages/@webex/plugin-cc/src/services/config/types.ts:114]: agent id we can pass from config

File: packages/@webex/plugin-cc/src/features/Agent.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import {LoginOption, WebexSDK} from '../types';
```
Comment (rsarika) [inline packages/@webex/plugin-cc/src/features/Agent.ts]: remove this file completely

File: packages/@webex/plugin-cc/src/cc.ts [hunk 7]
```diff
@@ -145,4 +146,23 @@ export default class ContactCenter extends WebexPlugin implements IContactCenter
...
   }
+
+  /**
+   * This is used for setting agent state.
+   * @param options
+   * @returns Promise<SetStateResponse>
+   * @throws Error
+   */
+
+  public async setAgentStatus(data: StateChange): Promise<SetStateResponse> {
```
Comment (rsarika) [inline packages/@webex/plugin-cc/src/cc.ts]: be consistant with naming. either state or status.

File: packages/@webex/plugin-cc/src/features/Agent.ts (hunk 1 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: This file is removed

File: packages/@webex/plugin-cc/src/features/Agent.ts (hunk 2 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: This file is removed

File: packages/@webex/plugin-cc/src/features/Agent.ts (hunk 3 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: This file is removed

File: packages/@webex/plugin-cc/src/features/Agent.ts (hunk 6 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/features/Agent.ts]: This file is removed

File: packages/@webex/plugin-cc/src/cc.ts (hunk 7 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/cc.ts]: Changed to `setAgentState`.

File: packages/@webex/plugin-cc/src/services/config/types.ts (hunk 5 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/config/types.ts:114]: Need a bit more info here

File: packages/@webex/plugin-cc/src/services/config/types.ts (hunk 4 above)
Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/config/types.ts:110]: Any reason we can't use an interface here?

Comment (rarajes2) [reply packages/@webex/plugin-cc/src/services/config/types.ts:110]: We generally use `type` where it contains just one type or a combination of multiple interfaces and types.

**Comment by rsarika** [comment]
closing this PR as we opened new PR
https://github.com/webex/webex-js-sdk/pull/3970

File: packages/@webex/plugin-cc/src/services/config/types.ts (hunk 5 above)
Comment (rsarika) [reply packages/@webex/plugin-cc/src/services/config/types.ts:114]: fallback to `this.agentConfig.agentId}` default
        ```data: {...data, agentId: data.agentId || this.agentConfig.agentId},```

File: packages/@webex/plugin-cc/src/services/config/types.ts (hunk 4 above)
Comment (rsarika) [reply packages/@webex/plugin-cc/src/services/config/types.ts:110]: we use type instead of interface

**Pull request closed without merging**